transport/internal/tlsmux/conn_sniffer.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tlsmux
+
+import (
+	"bytes"
+	"net"
+)
+
+// connSniffer wraps the connection and enables tlsmux to sniff inital bytes from the
+// connection efficiently.
+type connSniffer struct {
+	net.Conn
+
+	// set to true when sniffing mode is disabled.
+	disableSniffing bool
+	// buf stores bytes read from the underlying connection when in sniffing
+	// mode. When sniffing mode is disabled, buffered bytes is returned.
+	buf bytes.Buffer
+}
+
+// Read returns bytes read from the underlying connection. When sniffing is
+// true, data read from the connection is stored in the buffer. When sniffing
+// mode is disabled, data is first read from the buffer and once the buffer is
+// empty the underlying connection is read.
+func (c *connSniffer) Read(b []byte) (int, error) {
```

Comment:
It seems like `c.Conn.Read` is destructive, i.e you'll get the first 5 bytes only once.
If that is true, then `*connSniffer.Read` should `defer c.stopSniffing()` at some point, to avoid being misused by the caller.

Reply:
As this is a private sniffer, IMO, when to stop sniffing must be controlled by the muxer as it holds the sequence of sniffing - parse TLS header and stop sniffing. The number of bytes read (5 bytes) must be unknown to the sniffer.